src/tree/index.tsx

```diff
@@ -0,0 +1,415 @@
+			nodes.push(node);
+			if (!controlledExpandedNodes) {
+				if (expandedNodes.indexOf(node.id) !== -1) {
+					nodes = [...nodes, ...createNodeFlatMap(node.id)];
+				}
+			} else {
+				if (controlledExpandedNodes.indexOf(node.id) !== -1) {
+					nodes = [...nodes, ...createNodeFlatMap(node.id)];
+				}
+			}
+		});
+		return nodes;
+	}
+
+	function onKeyDown(event: KeyboardEvent) {
+		event.stopPropagation();
+		const nodes = createNodeFlatMap();
+		const activeIndex = nodes.findIndex((node) => node.id === activeNode);
+		const activeExpandedNodes = controlledExpandedNodes || expandedNodes;
+
+		switch (event.which) {
+			// select
+			case Keys.Enter:
+			case Keys.Space:
+				event.preventDefault();
+				if (activeNode && selectedNode !== activeNode && !nodes[activeIndex].hasChildren) {
+					selectable && selectNode(activeNode);
+					checkable && checkNode(activeNode, !checkedNodes.includes(activeNode));
+				}
+				break;
+
+			// next
+			case Keys.Down:
+				event.preventDefault();
+				activateNode(nodes[(activeIndex + 1) % nodes.length].id);
+				break;
+
+			// previous
+			case Keys.Up:
+				event.preventDefault();
+				if (activeIndex - 1 < 0) {
+					activateNode(nodes[nodes.length - 1].id);
+				} else {
+					activateNode(nodes[activeIndex - 1].id);
+				}
+				break;
+
+			// expand
+			case Keys.Right:
+				event.preventDefault();
+				if (
+					activeNode &&
+					!activeExpandedNodes.includes(activeNode) &&
+					nodes[activeIndex].hasChildren
+				) {
+					expandNode(activeNode);
+				}
+				break;
+
+			// collapse
+			case Keys.Left:
+				event.preventDefault();
+				if (activeNode && activeExpandedNodes.includes(activeNode)) {
+					collapseNode(activeNode);
+				}
+				break;
+		}
+	}
+
+	function mapNodeTree(nodeId: string = 'root') {
+		const options = createOptions(nodeId);
+		const info = meta(template, options({ query: { parent: nodeId } }), true);
+
+		if (info === undefined || isLoading(template, options())) {
+			return <LoadingIndicator />;
+		}
+
+		const results = getOrRead(
+			template,
+			options({ query: { parent: nodeId }, size: info.total })
+		);
+
+		const nodes = flat(results);
+
+		return (
+			<ol
+				classes={[
+					nodeId === 'root' ? classes.root : null,
+					classes.nodeParent,
+					theme.variant()
+				]}
+				focus={() => shouldFocus}
+				onkeydown={onKeyDown}
+				tabIndex={0}
+			>
+				{nodes.map((node) => {
+					const isExpanded = controlledExpandedNodes
+						? controlledExpandedNodes.indexOf(node.id) !== -1
+						: expandedNodes.indexOf(node.id) !== -1;
+					return (
+						<li
+							classes={[
+								classes.node,
+								node.hasChildren && classes.leaf,
+								selectable && classes.selectable,
+								node.id === selectedNode && classes.selected
+							]}
+						>
+							<TreeNode
+								activeNode={activeNode}
+								checkable={checkable}
+								selectable={selectable}
+								checkedNodes={controlledCheckedNodes || checkedNodes}
```

Comment:
I thought you were changing the `TreeNodeProperties` to simply be passed a `boolean` for `checked` / `selected` / `disabled` etc? is there a reason why it needs to know about all of the checked nodes etc rather than just about itself?

Reply:
I've updated these to be `booleans`